### PR TITLE
Kubernetes: bring kube-dns implementation up-to-date

### LINKED ISSUE #2999: Pods are unable to resolve DNS for services both internally and externally.
**Is this a request for help?**:
Yes
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
Issue
---

**What version of acs-engine?**:
0.15.2
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes Version: 1.10.1

**What happened**:
All our internal services report a DNS resolution failure when called, results in an overall network outage in our cluster.

**What you expected to happen**:
DNS should be resolved. 

**How to reproduce it** (as minimally and precisely as possible):
Happens erratically. Root cause of the problem is still unknown. 

**Anything else we need to know**:

We are facing this issue in "production". It results in a huge business impact as all services running on our cluster goes down. 

Observation: 

1. Calls to external services fail. 
2. Inter service calls within the cluster fail.
3. Due to the above - some pods in the cluster to go CrashLoopBackOff mode. 
 
Current Hack/Fix that we use today which is temporary:

1. Remove the failing pods which go in CrashLoopbackOff state.
2. Restart Kubedns pods present in the cluster.
3. Restart all worker nodes in the cluster - a very dangerous operation.

Attached Kubedns logs

```
⇒  kubectl logs -n kube-system kube-dns-v20-59b4f7dc55-4dxh8 kubedns
I0515 12:30:47.752415       1 dns.go:48] version: 1.14.8
I0515 12:30:47.753947       1 server.go:71] Using configuration read from directory: /kube-dns-config with period 10s
I0515 12:30:47.754135       1 server.go:119] FLAG: --alsologtostderr="false"
I0515 12:30:47.754177       1 server.go:119] FLAG: --config-dir="/kube-dns-config"
I0515 12:30:47.754186       1 server.go:119] FLAG: --config-map=""
I0515 12:30:47.754192       1 server.go:119] FLAG: --config-map-namespace="kube-system"
I0515 12:30:47.754197       1 server.go:119] FLAG: --config-period="10s"
I0515 12:30:47.754204       1 server.go:119] FLAG: --dns-bind-address="0.0.0.0"
I0515 12:30:47.754209       1 server.go:119] FLAG: --dns-port="10053"
I0515 12:30:47.754216       1 server.go:119] FLAG: --domain="cluster.local."
I0515 12:30:47.754223       1 server.go:119] FLAG: --federations=""
I0515 12:30:47.754229       1 server.go:119] FLAG: --healthz-port="8081"
I0515 12:30:47.754235       1 server.go:119] FLAG: --initial-sync-timeout="1m0s"
I0515 12:30:47.754241       1 server.go:119] FLAG: --kube-master-url=""
I0515 12:30:47.754247       1 server.go:119] FLAG: --kubecfg-file=""
I0515 12:30:47.754253       1 server.go:119] FLAG: --log-backtrace-at=":0"
I0515 12:30:47.754261       1 server.go:119] FLAG: --log-dir=""
I0515 12:30:47.754267       1 server.go:119] FLAG: --log-flush-frequency="5s"
I0515 12:30:47.754272       1 server.go:119] FLAG: --logtostderr="true"
I0515 12:30:47.754277       1 server.go:119] FLAG: --nameservers=""
I0515 12:30:47.754291       1 server.go:119] FLAG: --stderrthreshold="2"
I0515 12:30:47.754297       1 server.go:119] FLAG: --v="2"
I0515 12:30:47.754302       1 server.go:119] FLAG: --version="false"
I0515 12:30:47.754310       1 server.go:119] FLAG: --vmodule=""
I0515 12:30:47.754411       1 server.go:201] Starting SkyDNS server (0.0.0.0:10053)
I0515 12:30:47.754470       1 server.go:222] Skydns metrics not enabled
I0515 12:30:47.754480       1 dns.go:146] Starting endpointsController
I0515 12:30:47.754485       1 dns.go:149] Starting serviceController
I0515 12:30:47.754528       1 logs.go:41] skydns: ready for queries on cluster.local. for tcp://0.0.0.0:10053 [rcache 0]
I0515 12:30:47.754547       1 logs.go:41] skydns: ready for queries on cluster.local. for udp://0.0.0.0:10053 [rcache 0]
I0515 12:30:48.254763       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:48.754798       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:49.254807       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:49.754810       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:50.254806       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:50.754734       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:51.254743       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:51.754780       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:52.254785       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:52.754699       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:52.891031       1 logs.go:41] skydns: failure to forward request "read udp 10.241.5.12:44073->168.63.129.16:53: i/o timeout"
I0515 12:30:53.254755       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
I0515 12:30:53.754780       1 dns.go:173] Waiting for services and endpoints to be initialized from apiserver...
```

### LINKED ISSUE #3534: kube-dns-v20 deployment
**Is this a request for help?**:

Yes.

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

Issue.

**What version of acs-engine?**:

v0.14.5 & v0.20.0

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
acs-engine 0.14.5 | k8s 1.9.6
acs-engine 0.20.0 | k8s 1.11.0

**What happened**:
I'm running into an issue that's pretty similar to https://github.com/Azure/acs-engine/issues/2999 where the kube-dns-v20 deployments go into a CrashBackoffLoop **after** adding a config map with my custom consul upstream server. 

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: kube-dns
  namespace: kube-system
data:
  upstreamNameservers: |
    ["<consul box ip"]
```
It doesn't happen right away, but the problem usually presents itself after an hr or so and starts with not being able to do nslookups on the internal kubernetes network, causing the DNS stack to loop since the health check fails.

```
kubectl logs -n kube-system kube-dns-v20-77b8c6b4c-zz492 -c healthz -f
2018/07/23 22:31:20 Healthz probe on /healthz-dnsmasq error: Result of last exec: nslookup: can't resolve 'kubernetes.default.svc.cluster.local'
, at 2018-07-23 22:30:51.167132151 +0000 UTC, error exit status 1
2018/07/23 22:31:30 Healthz probe on /healthz-dnsmasq error: Result of last exec: nslookup: can't resolve 'kubernetes.default.svc.cluster.local'
, at 2018-07-23 22:31:25.407528227 +0000 UTC, error exit status 1
2018/07/23 22:31:40 Healthz probe on /healthz-dnsmasq error: Result of last exec: nslookup: can't resolve 'kubernetes.default.svc.cluster.local'
, at 2018-07-23 22:31:25.407528227 +0000 UTC, error exit status 1
2018/07/23 22:31:50 Healthz probe on /healthz-dnsmasq error: Result of last exec: nslookup: can't resolve 'kubernetes.default.svc.cluster.local'
, at 2018-07-23 22:31:25.407528227 +0000 UTC, error exit status 1
2018/07/23 22:32:00 Healthz probe on /healthz-dnsmasq error: Result of last exec: nslookup: can't resolve 'kubernetes.default.svc.cluster.local'
, at 2018-07-23 22:31:58.678927868 +0000 UTC, error exit status 1
```

Though internal lookups fail, external lookups still work when the cluster comes up momentarily. In testing with minikube, I noticed that the args for dnsmasq are slightly different.  My minikube cluster's dns stack includes these args that are not present in azure: 

```
        - --no-negcache
        - --server=/cluster.local/127.0.0.1#10053
```
After deleting the DNS stack and recreating them with those new values added, my DNS pods have been stable for about a day. Is there a reason why those args aren't included in the kube-dns stack that's spun up via azure's images? I'm not sure if this is the fix, but can those be added?

kube-dns:1.14.8 is missing the same args but has a slightly different issue where requests just don't route to the upstreamserver at all from the pods for external services, which is weird and isn't how it's supposed to work according to kubernetes documentation, thus leading me to try 1.14.10 instead.

Would appreciate any insight...

**What you expected to happen**:
DNS cluster should be stable after adding custom upstream.

**How to reproduce it** (as minimally and precisely as possible):
Create a cluster with k8s 1.11.0, add an upstreamNameserver via configmap, resolve a few things and wait about an hr for the DNS stack to start crash looping.  

**Anything else we need to know**:
Our consul DNS is set up with google's dns as its recursors for anything outside of the consul domain.
kube components that we're using:
k8s-gcrio.azureedge.net/exechealthz-amd64:1.2
k8s-gcrio.azureedge.net/k8s-dns-dnsmasq-nanny-amd64:1.14.10
k8s-gcrio.azureedge.net/k8s-dns-kube-dns-amd64:1.14.10

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**: Basing kube-dns implementation on the kubernetes-recommended base configs. 

Primary change is to replace exechealthz with sidecar. Additionally added `--no-negcache` config to dnsmasq.

As a reference, here are the base configs.

For v1.11:

https://github.com/kubernetes/kubernetes/blob/v1.11.0/cluster/addons/dns/kube-dns/kube-dns.yaml.base

For v1.10:

https://github.com/kubernetes/kubernetes/blob/v1.10.0/cluster/addons/dns/kube-dns.yaml.base

For v1.9:

https://github.com/kubernetes/kubernetes/blob/v1.9.0/cluster/addons/dns/kube-dns.yaml.base

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #3534 fixes #2999

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```
use exechealthz v1.3.0 in k8s 1.11
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
@feiskyer @andyzhangx FYI, testing new exechealthz on v1.11 clusters. Any reasons that you're aware of to backport this update for earlier cluster versions? (Or are there any reasons why not to move forward w/ new exechealthz version?)

**Comment by feiskyer** [comment]
@jackfrancis kube-dns has moved to sidecar container for health checking since https://github.com/kubernetes/kubernetes/pull/38992 (v.1.6.0). I suggest we also use it (e.g. k8s.gcr.io/k8s-dns-sidecar-amd64:1.14.10) for our cluster.

**Comment by jackfrancis** [comment]
Reference for v1.11 kube-dns changes:

https://github.com/kubernetes/kubernetes/blob/v1.11.0/cluster/addons/dns/kube-dns/kube-dns.yaml.base

File: parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 1]
```diff
@@ -146,6 +146,8 @@ spec:
         - mountPath: /kube-dns-config
           name: kube-dns-config
       - args:
```
Comment (feiskyer) [inline parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml]: healthz container should be removed, as sidecar container has been added below.

Comment (feiskyer) [reply parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml]: Also deployed with https://github.com/kubernetes/kubernetes/blob/v1.11.0/cluster/addons/dns/kube-dns/kube-dns.yaml.base, it works well with acs-engine deployed cluster (v1.11.0).

**Comment by jackfrancis** [comment]
@feiskyer Thanks for the continued guidance! I've converted the 1.11 kube-dns implementation here to follow the base kubernetes example. Initial tests suggest that HPA is failing, but everything else in our test surface area checks out.

Does anything look suspicious here that would break HPA?

https://github.com/jackfrancis/acs-engine/blob/1bcf1af22bc7cd446c95766fec93b6f05930d921/parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml

**Comment by feiskyer** [comment]
@jackfrancis Did you mean HPA for other pods or dns-horizontal-autoscaler for kube-dns? The change doesn't seem related with HPA for other pods.

**Comment by jackfrancis** [comment]
@feisker just regular HPA. We test deploying nginx, attaching HPA config to it, and then hitting with load.

**Comment by feiskyer** [comment]
Have you checked HPA events? It should include some hints

**Comment by jackfrancis** [comment]
@feiskyer so, a follow-up run had different side-effects.

I observed `kube-dns` pod go from `CrashLoopBackOff` to `Error`.

See:

```
$ kubectl logs kube-dns-7f9df74d5b-6g7g2 -n kube-system -c dnsmasq
I0726 19:42:19.211202       1 main.go:74] opts: {{/usr/sbin/dnsmasq [-k --cache-size=1000 --no-negcache --log-facility=- --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053] true} /etc/k8s/dns/dnsmasq-nanny 10000000000}
I0726 19:42:19.211373       1 nanny.go:94] Starting dnsmasq [-k --cache-size=1000 --no-negcache --log-facility=- --server=/in-addr.arpa/127.0.0.1#10053 --server=/ip6.arpa/127.0.0.1#10053]
I0726 19:42:19.547769       1 nanny.go:119] 
W0726 19:42:19.547801       1 nanny.go:120] Got EOF from stdout
I0726 19:42:19.547815       1 nanny.go:116] dnsmasq[9]: started, version 2.78 cachesize 1000
I0726 19:42:19.547826       1 nanny.go:116] dnsmasq[9]: compile time options: IPv6 GNU-getopt no-DBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-DNSSEC loop-detect inotify
I0726 19:42:19.547831       1 nanny.go:116] dnsmasq[9]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0726 19:42:19.547835       1 nanny.go:116] dnsmasq[9]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0726 19:42:19.547945       1 nanny.go:116] dnsmasq[9]: reading /etc/resolv.conf
I0726 19:42:19.547954       1 nanny.go:116] dnsmasq[9]: using nameserver 127.0.0.1#10053 for domain ip6.arpa 
I0726 19:42:19.547958       1 nanny.go:116] dnsmasq[9]: using nameserver 127.0.0.1#10053 for domain in-addr.arpa 
I0726 19:42:19.547961       1 nanny.go:116] dnsmasq[9]: using nameserver 168.63.129.16#53
I0726 19:42:19.548104       1 nanny.go:116] dnsmasq[9]: read /etc/hosts - 7 addresses
```

and:

```
$ kubectl logs kube-dns-7f9df74d5b-6g7g2 -n kube-system -c sidecar
I0726 19:40:02.174431       1 main.go:51] Version v1.14.8.3
I0726 19:40:02.174499       1 server.go:45] Starting server (options {DnsMasqPort:53 DnsMasqAddr:127.0.0.1 DnsMasqPollIntervalMs:5000 Probes:[{Label:kubedns Server:127.0.0.1:10053 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:33} {Label:dnsmasq Server:127.0.0.1:53 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:33}] PrometheusAddr:0.0.0.0 PrometheusPort:10054 PrometheusPath:/metrics PrometheusNamespace:kubedns})
I0726 19:40:02.174527       1 dnsprobe.go:75] Starting dnsProbe {Label:kubedns Server:127.0.0.1:10053 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:33}
I0726 19:40:02.174577       1 dnsprobe.go:75] Starting dnsProbe {Label:dnsmasq Server:127.0.0.1:53 Name:kubernetes.default.svc.cluster.local. Interval:5s Type:33}
W0726 19:40:02.174835       1 server.go:64] Error getting metrics from dnsmasq: read udp 127.0.0.1:56325->127.0.0.1:53: read: connection refused
```

File: parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 2]
```diff
@@ -104,75 +141,72 @@ spec:
...
+        args:
+        - -v=2
+        - -logtostderr
+        - -configDir=/etc/k8s/dns/dnsmasq-nanny
+        - -restartDnsmasq=true
+        - --
+        - -k
+        - --cache-size=1000
+        - --no-negcache
+        - --log-facility=-
```
Comment (feiskyer) [inline parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml]: missing `--server=/cluster.local/127.0.0.1#10053` here?

File: parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 3]
```diff
@@ -25,75 +36,101 @@ spec:
...
-        scheduler.alpha.kubernetes.io/critical-pod: ""
-        prometheus.io/scrape: "true"
-        prometheus.io/port: "10055"
       labels:
         k8s-app: kube-dns
-        kubernetes.io/cluster-service: "true"
-        version: v20
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+        seccomp.security.alpha.kubernetes.io/pod: 'docker/default'
```
Comment (feiskyer) [inline parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml]: I don't think we need this either, we may remove this annotation

**Comment by feiskyer** [comment]
It works fine on my cluster after adding `--server=/cluster.local/127.0.0.1#10053`.

**Comment by jackfrancis** [comment]
That worked @feiskyer, thank you! The changes here are only for 1.11. Do you recommend doing a similar kube-dns conversion for any other k8s versions?

**Comment by jackfrancis** [comment]
@feiskyer for your review. I audited the k8s codebase and determined that the sidecar implementation has been in place since 1.7 (at least, I didn't check earlier).

To be conservative, and with the prediction that there are more clusters on 1.8 and below in the wild (1.8 is the default when none is provided), I have converted 1.9 and above to adhere to an implementation that looks like the upstream base example.

Thoughts? Any reservations about these changes? Thanks again for your eyes!

**Comment by feiskyer** [comment]
> I have converted 1.9 and above to adhere to an implementation that looks like the upstream base example.

I'm also suggesting this, so we're consistent with upstream.

**Pull request merged**